File: src/main/java/com/github/iamniklas/liocore/procedures/variants/JsonProcedure.java
package com.github.iamniklas.liocore.procedures.variants;

import com.github.iamniklas.liocore.led.LEDDataBundle;
import com.github.iamniklas.liocore.led.colorspace.LIOColor;
import com.github.iamniklas.liocore.led.json.LEDJsonProcedure;
import com.github.iamniklas.liocore.led.json.interpreter.FileVersions;
import com.github.iamniklas.liocore.led.json.interpreter.LEDInterpreter;
import com.github.iamniklas.liocore.network.LEDUpdateModel;
import com.github.iamniklas.liocore.procedures.Procedure;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;

public class JsonProcedure extends Procedure {

    private LEDDataBundle bundle;

    private LEDJsonProcedure ledJsonProcedure;

    public JsonProcedure(LEDUpdateModel _ledUpdateModel) {
        super(_ledUpdateModel);
        bundle = _ledUpdateModel.bundle;

        ledJsonProcedure = loadFromFile(bundle.path);
    }

    @Override
    public void start() {
        super.start();
    }

    @Override
    public void update() {
        for (int i = 0; i < ledJsonProcedure.metaInfo.ledCount; i++) {
            strip.setPixel(i, LIOColor.fromRGB(ledJsonProcedure.ledStates[step].ledState[i]));
        }
        step += 5;
        if(step >= steps) {
            finishProcedure();
        }
    }

    LEDJsonProcedure loadFromFile(String _path) {
        byte[] encoded;
        try {
            encoded = Files.readAllBytes(Paths.get(_path));
            return LEDInterpreter.interpretJson(FileVersions.Latest, new String(encoded, StandardCharsets.UTF_8));
        }
        catch (IOException ioe) {
            ioe.printStackTrace();
            return null;
        }
    }

    @Override
    public void updateLEDDataBundle(LEDDataBundle ledDataBundle) {

    }
}

File: src/main/java/com/github/iamniklas/liocore/procedures/JsonProcedure.kt
package com.github.iamniklas.liocore.procedures

import com.github.iamniklas.liocore.led.LEDDataBundle
import com.github.iamniklas.liocore.led.json.LEDJsonProcedure
import com.github.iamniklas.liocore.led.json.interpreter.FileVersions
import com.github.iamniklas.liocore.led.json.interpreter.LEDInterpreter
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths

class JsonProcedure(_bundle: LEDDataBundle) : Procedure(_bundle) {
    var mLEDJsonProcedure: LEDJsonProcedure?
    override fun start() {}
    public override fun update() {
        for (i in 0 .. mLEDJsonProcedure!!.mMetaInfo!!.mLedCount) {
            mStrip!!.setPixel(i, mLEDJsonProcedure!!.mLEDStates[mStep]!!.mLEDState[i].toSystemColor())
        }
        mStep += 5
        if (mStep >= mSteps) {
            finishProcedure()
        }
    }

    private fun loadFromFile(_path: String?): LEDJsonProcedure? {
        val encoded: ByteArray
        return try {
            encoded = Files.readAllBytes(Paths.get(_path))
            LEDInterpreter.interpretJson(FileVersions.V0_0_1, "")//String(encoded, StandardCharsets.US_ASCII))
        } catch (e: IOException) {
            e.printStackTrace()
            null
        }
    }

    init {
        mLEDJsonProcedure = loadFromFile(_bundle.path)
        mSteps = mLEDJsonProcedure!!.mLEDStates.size - 1

        //mIsSubProcedure = (boolean) _bundle.get(ProcedureBundleFields.IS_SUB_PROCEDURE);
    }
}